motion_computation/src/psm_to_external_object_convertor.cpp

```diff
@@ -0,0 +1,501 @@
+
+  if (fabs(map_point.x()) > 10000.0 ||
+      fabs(map_point.y()) > 10000.0) {  // Above 10km from map origin earth curvature will start to have a negative
+                                        // impact on system performance
+
+    RCLCPP_WARN_STREAM(rclcpp::get_logger("motion_computation::conversion"), 
+        "Distance from map origin is larger than supported by system "
+        "assumptions. Strongly advise "
+        "alternative map origin be used. ");
+  }
+
+  //// Convert the orientation information into the map frame
+  // This logic assumes that the orientation difference between an NED frame
+  // located at the map origin and an NED frame located at the GNSS point are
+  // sufficiently small that they can be ignored. Therefore it is assumed the
+  // heading report of the GNSS system regardless of its postion in the map
+  // without change in its orientation will give the same result (as far as we
+  // are concered).
+
+  tf2::Quaternion R_m_n(ned_in_map_rotation);  // Rotation of NED frame in map
+                                               // frame
+  tf2::Quaternion R_n_h;                       // Rotation of sensor heading report in NED frame
+  R_n_h.setRPY(0, 0, heading * wgs84_utils::DEG2RAD);
+
+  tf2::Quaternion R_m_s = R_m_n * R_n_h;  // Rotation of sensor in map frame under assumption that
+                                          // distance from map origin is sufficiently small so as to
+                                          // ignore local changes in NED orientation
+
+  RCLCPP_DEBUG_STREAM(rclcpp::get_logger("motion_computation::conversion"), "R_m_n (x,y,z,w) : ( " << R_m_n.x() << ", " << R_m_n.y() << ", " << R_m_n.z() << ", " << R_m_n.w());
+
+  RCLCPP_DEBUG_STREAM(rclcpp::get_logger("motion_computation::conversion"), "R_n_h (x,y,z,w) : ( " << R_n_h.x() << ", " << R_n_h.y() << ", " << R_n_h.z() << ", " << R_n_h.w());
+
+  RCLCPP_DEBUG_STREAM(rclcpp::get_logger("motion_computation::conversion"), "R_m_s (x,y,z,w) : ( " << R_m_s.x() << ", " << R_m_s.y() << ", " << R_m_s.z() << ", " << R_m_s.w());
+
+  tf2::Transform T_m_s(R_m_s, tf2::Vector3(map_point.x(), map_point.y(),
+                                           map_point.z()));  // Reported position and orientation
+                                                             // of sensor frame in map frame
+
+        tf2::Transform T_m_n_no_heading(R_m_n, tf2::Vector3(0, 0, 0)); // Used to transform the covariance
+
+        // This covariance represents the covariance of the NED frame N-lat,
+        // E-lon, heading- angle east of north This means that the covariance is
+        // in the NED frame, and needs to be transformed to the map frame
+        // clang-format off
+        std::array<double, 36> input_covariance = { 
+          lat_variance, 0, 0,  0, 0, 0,
+          0, lon_variance, 0,  0, 0, 0,
+          0, 0, 0,  0, 0, 0,
```

Comment:
I probably need a 1 in the third element here to keep the diagonal populated

Reply:
In that case, do we also need to look into cases where the variance values are 0? There is a case defined where that happens (Line 159)

Reply:
Hmm that's a fair point. Maybe set them to 1 as well if that happens.